source/common/quic/envoy_tls_server_handshaker.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "envoy/ssl/private_key/private_key_callbacks.h"
+
+#include "source/common/quic/quic_server_transport_socket_factory.h"
+
+#include "quiche/quic/core/crypto/proof_source.h"
+#include "quiche/quic/core/tls_server_handshaker.h"
+
+namespace Envoy {
+namespace Quic {
+
+class EnvoyTlsServerHandshaker : public quic::TlsServerHandshaker,
+                                 public Envoy::Ssl::PrivateKeyConnectionCallbacks {
+public:
+  EnvoyTlsServerHandshaker(quic::QuicSession* session,
+                           const quic::QuicCryptoServerConfig* crypto_config,
+                           Envoy::Event::Dispatcher& dispatcher);
+
+  void onPrivateKeyMethodComplete() override;
+
+  Envoy::Event::Dispatcher& dispatcher() { return dispatcher_; }
+
+protected:
+  void FinishHandshake() override;
+  std::unique_ptr<quic::ProofSourceHandle> MaybeCreateProofSourceHandle() override;
+
+private:
+  // DefaultProofSourceHandle delegates all operations to the shared proof
+  // source.
+  class EnvoyDefaultProofSourceHandle : public quic::ProofSourceHandle,
```

Comment:
Not sure we should do my own cert selection. The cert selection should be same whatever the private key provider used or not. And I don't think I have enough data to do the cert selection. Look at the EnvoyQuicProofSource, I need the filter_manager, listen socket etc to do the cert selection. I can't get them in the proof source handle.

Reply:
If you pass down the transport_socket_factory to the handle, it should have cert chain or the configuration of cert chain. And you won't need filter_manager, listen socket, proof_source, etc, as the transport socket_factory is retrieved from proof source using those properties earlier during connection creation. The retrieval logic is pretty much similar to EnvoyQuicProofSource. The shared proof source needs to retrieve the filter chain with transport socket config again because it doesn't have the per-connection states there, i.e. the matched filter chain. But your handle should have those states to begin with.

Reply:
yea, that's ture, I can get the transport_socket_factory. let me give try. Then let us skip all the shared proof source, If the cert is configured with private key provider, then invoke the provider in the proof source handle directly. Is it same with what you expected?

Reply:
Exactly. You may need to do some refactoring in QuicDownStreamTransportSocketFactory to expose a cert selection interface to share between proof source and here.